### PR TITLE
Waveforms view mode - geom layout or vertical

### LINKED ISSUE #318: Waveforms view mode - geom layout or not

### PR DESCRIPTION
**Pull request by magland**
Fixes #318

### CODE REVIEW

File: src/extensions/averagewaveforms/AverageWaveformsView/AverageWaveformView.tsx [hunk 1]
```diff
@@ -31,6 +31,7 @@ const AverageWaveformView: FunctionComponent<Props> = ({ boxSize, plotData, args
     return (
         <WaveformWidget
             waveform={plotData.average_waveform}
+            layoutMode={selection.waveformsMode || 'geom'}
```
Comment (magland) [inline src/extensions/averagewaveforms/AverageWaveformsView/AverageWaveformView.tsx:34]: layoutMode tells whether to use elec geom to display the waveforms. These screenshots demonstrate the two alternatives:

image

image

File: src/extensions/averagewaveforms/AverageWaveformsView/electrodesLayer.ts [hunk 2]
```diff
@@ -124,8 +124,14 @@ export const createElectrodesLayer = () => {
                                 : hovered
                                     ? colors.hover
                                     : colors.base
-            painter.fillEllipse(e.rect, {color: color})
-            painter.drawEllipse(e.rect, {color: colors.border})
+            const layoutMode = props.layoutMode
+            if (layoutMode === 'geom') {
```
Comment (magland) [inline src/extensions/averagewaveforms/AverageWaveformsView/electrodesLayer.ts:128]: Either draw ellipse or a line, depending on the layout mode

File: src/extensions/averagewaveforms/AverageWaveformsView/setupElectrodes.ts [hunk 3]
```diff
@@ -44,8 +44,61 @@ const getElectrodesBoundingBox = (electrodeLocations: number[][], radius: number
...
-const setupElectrodes = (args: {width: number, height: number, electrodeLocations: Vec2[], electrodeIds: number[]}) => {
-    const { width, height, electrodeLocations, electrodeIds } = args
+export const getElectrodesAspectRatio = (electrodeLocations: Vec2[]) => {
+    const radius = computeRadius(electrodeLocations)
+    let boundingBox = getElectrodesBoundingBox(electrodeLocations, radius)
+    let boxAspect = getWidth(boundingBox) / getHeight(boundingBox)
+    return boxAspect
+}
+
+const setupVerticalElectrodes = ({width, height, electrodeIds}: {width: number, height: number, electrodeIds: number[]}) => {
```
Comment (magland) [inline src/extensions/averagewaveforms/AverageWaveformsView/setupElectrodes.ts:54]: use setupVerticalElectrodes for the 'vertical' layout mode

File: src/extensions/mountainview/MVSortingView/MVSortingView.tsx [hunk 4]
```diff
@@ -212,6 +213,11 @@ const MVSortingView: FunctionComponent<SortingViewProps> = (props) => {
                             selectionDispatch={props.selectionDispatch}
                         />
                     </Expandable>
+
+                    {/* Options */}
```
Comment (magland) [inline src/extensions/mountainview/MVSortingView/MVSortingView.tsx:217]: A new options box in the left control panel allowing toggle between geom and vertical modes

File: src/extensions/snippets/SnippetsView/SnippetsRow.tsx [hunk 5]
```diff
@@ -166,35 +167,37 @@ const filterSnippetVisibleElectrodes = (s: Snippet, electrodeIds: number[] | und
 
 const SnippetsRow: FunctionComponent<Props> = ({ recording, sorting, selection, selectionDispatch, unitId, height, noiseLevel }) => {
     const {snippets, info} = useSnippets({recording, sorting, visibleElectrodeIds: selection.visibleElectrodeIds, unitId, timeRange: selection.timeRange || null})
+    const electrodeLocations = info?.channel_locations
+    const boxWidth = useMemo(() => {
```
Comment (magland) [inline src/extensions/snippets/SnippetsView/SnippetsRow.tsx:171]: Determine the box width based on the aspect ratio of the electrodes

File: src/extensions/averagewaveforms/AverageWaveformsView/AverageWaveformView.tsx [hunk 6]
```diff
@@ -31,6 +31,7 @@ const AverageWaveformView: FunctionComponent<Props> = ({ boxSize, plotData, args
     return (
```
Comment (jsoules) [inline src/extensions/averagewaveforms/AverageWaveformsView/AverageWaveformView.tsx:31]: I'm a little hazy on the distinction between `AverageWaveformView.tsx` and `AverageWaveformView2.tsx`--I thought the former was deprecated, but I see that's not the case. What's the difference?

File: src/extensions/averagewaveforms/AverageWaveformsView/electrodesLayer.ts [hunk 7]
```diff
@@ -124,8 +124,14 @@ export const createElectrodesLayer = () => {
...
+                painter.fillEllipse(e.rect, {color: color})
+                painter.drawEllipse(e.rect, {color: colors.border})
+            }
+            else if (layoutMode === 'vertical') {
+                painter.drawLine(e.rect.xmin, (e.rect.ymin + e.rect.ymax) / 2, e.rect.xmax, (e.rect.ymin + e.rect.ymax) / 2, {color: colors.border})
```
Comment (jsoules) [inline src/extensions/averagewaveforms/AverageWaveformsView/electrodesLayer.ts:133]: Draws a midpoint-line that the waveform will display above and below

File: src/extensions/averagewaveforms/AverageWaveformsView/setupElectrodes.ts [hunk 8]
```diff
@@ -44,8 +44,61 @@ const getElectrodesBoundingBox = (electrodeLocations: number[][], radius: number
     }
 }
 
-const setupElectrodes = (args: {width: number, height: number, electrodeLocations: Vec2[], electrodeIds: number[]}) => {
-    const { width, height, electrodeLocations, electrodeIds } = args
+export const getElectrodesAspectRatio = (electrodeLocations: Vec2[]) => {
+    const radius = computeRadius(electrodeLocations)
+    let boundingBox = getElectrodesBoundingBox(electrodeLocations, radius)
+    let boxAspect = getWidth(boundingBox) / getHeight(boundingBox)
```
Comment (jsoules) [inline src/extensions/averagewaveforms/AverageWaveformsView/setupElectrodes.ts:50]: can both be `const` in this context

File: src/extensions/averagewaveforms/AverageWaveformsView/setupElectrodes.ts [hunk 9]
```diff
@@ -44,8 +44,61 @@ const getElectrodesBoundingBox = (electrodeLocations: number[][], radius: number
...
+        })
+        return {
+            label: eid + '',
+            id: eid,
+            x: 0.5,
+            y,
+            rect,
+            transform: transform0
+        }
+    })
```
Comment (jsoules) [inline src/extensions/averagewaveforms/AverageWaveformsView/setupElectrodes.ts:81]: This function is trying to create bounding canvas areas for each waveform for the vertical display? With each channel mapped to its own space, whose width is the full width of the unit and whose height is 1/nth of it, less a margin?

The screenshot shows it works, but it's not super easy to follow what it does, nor match up the behavior to confirm that it does what's intended. We should add some comments here (at least l. 65-81).

File: src/extensions/averagewaveforms/AverageWaveformsView/AverageWaveformView.tsx (hunk 6 above)
Comment (magland) [reply src/extensions/averagewaveforms/AverageWaveformsView/AverageWaveformView.tsx:31]: It is deprecated. Accidentally modified it.

**Pull request merged**